openssl/src/crypto/pkcs7/pk7_smime.rs

```diff
@@ -0,0 +1,225 @@
+        const PKCS7_REUSE_DIGEST = ffi::PKCS7_REUSE_DIGEST;
+        const PKCS7_NO_DUAL_CONTENT = ffi::PKCS7_NO_DUAL_CONTENT;
+    }
+}
+
+impl PKCS7 {
+    pub fn smime_write(&self, input: Vec<u8>, flags: PKCS7Flags) -> Result<Vec<u8>, ErrorStack> {
+        unsafe {
+            let input_bio = MemBioSlice::new(input.as_slice())?;
+
+            let output = MemBio::new()?;
+
+            ffi::SMIME_write_PKCS7(output.as_ptr(), self.0, input_bio.as_ptr(), flags.bits);
+
+            Ok(output.get_buf().to_owned())
+        }
+    }
+
+    pub fn smime_read(input: Vec<u8>, bcount: Option<Vec<u8>>) -> Result<PKCS7, ErrorStack> {
+        let input_bio = MemBioSlice::new(input.as_slice())?;
+
+        match bcount {
+            Some(data) => {
+                let bcount_bio = MemBioSlice::new(data.as_slice())?;
+                Ok(PKCS7(unsafe { ffi::SMIME_read_PKCS7(input_bio.as_ptr(), &mut bcount_bio.as_ptr()) }))
+            },
+            None => Ok(PKCS7(unsafe { ffi::SMIME_read_PKCS7(input_bio.as_ptr(), null_mut()) }))
+        }
+    }
+
+    pub fn decrypt(&self, pkey: PKey, cert: X509) -> Result<Vec<u8>, ErrorStack> {
+        let output = MemBio::new()?;
+
+        unsafe {
+            ffi::PKCS7_decrypt(self.0, pkey.as_ptr(), cert.as_ptr(), output.as_ptr(), 0);
+
+            Ok(output.get_buf().to_owned())
+        }
+    }
+
+    pub fn encrypt(certs: Vec<X509>, input: Vec<u8>, cypher: Cipher, flags: PKCS7Flags) -> Result<PKCS7, ErrorStack> {
+        let input_bio = MemBioSlice::new(input.as_slice())?;
+        let certs_stack: Stack<X509> = certs
+            .iter()
+            .fold(Ok(Stack::new()?), | stack_result, cert | {
+                match stack_result {
+                    Ok(mut stack) => {
+                        stack.push(cert.clone())?;
+                        Ok(stack)
+                    },
+                    Err(error) => Err(error)
+                }
+            })?;
+
+        Ok(PKCS7(unsafe { ffi::PKCS7_encrypt(certs_stack.as_ptr(), input_bio.as_ptr(), cypher.as_ptr(), flags.bits) }))
+    }
+
+    pub fn sign(signcert: X509, pkey: PKey, certs: Vec<X509>, input: Vec<u8>, flags: PKCS7Flags) -> Result<PKCS7, ErrorStack> {
+        let input_bio = MemBioSlice::new(input.as_slice())?;
+        let certs_stack: Stack<X509> = certs
+            .iter()
+            .fold(Ok(Stack::new()?), | stack_result, cert | {
+                match stack_result {
+                    Ok(mut stack) => {
+                        stack.push(cert.clone())?;
+                        Ok(stack)
+                    },
+                    Err(error) => Err(error)
+                }
+            })?;
+
+        Ok(PKCS7(unsafe { ffi::PKCS7_sign(signcert.as_ptr(), pkey.as_ptr(), certs_stack.as_ptr(), input_bio.as_ptr(), flags.bits) }))
+    }
+
+    pub fn verify(&self, certs: Vec<X509>, store: X509Store, indata: Option<Vec<u8>>, out: &MemBio, flags: PKCS7Flags) -> Result<bool, ErrorStack> {
```

Comment:
`&StackRef<X509>`, `&X509StoreRef`, `Option<&[u8]>`.

`MemBio` isn't a publicly exposed type, so we'll probably want to pass in a `&mut Vec<u8>` instead. Maybe an `Option<&mut Vec<u8>>` since out data is optional?